athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/common/PartitionUtil.java

```diff
@@ -0,0 +1,195 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.amazonaws.athena.connectors.gcs.common;
+
+import com.amazonaws.services.glue.model.Column;
+import com.amazonaws.services.glue.model.Table;
+import org.apache.arrow.vector.complex.reader.FieldReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static com.amazonaws.athena.connectors.gcs.GcsConstants.PARTITION_PATTERN_KEY;
+
+public class PartitionUtil
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(PartitionUtil.class);
+
+    /**
+     * Partition pattern regular expression to be used in compiling Pattern
+     */
+    private static final String PARTITION_PATTERN_REGEX = "\\{(.*?)}";
+
+    /**
+     * Pattern from a regular expression that identifies a match in a phrases to see if there is any
+     * partition key variable placeholder. A partition key variable placeholder looks something like the following:
+     * year={year}/month={month}
+     * Here, {year} and {month} are the partition key variable placeholders
+     */
+    private static final Pattern PARTITION_PATTERN = Pattern.compile(PARTITION_PATTERN_REGEX);
+
+    /**
+     * Match any alpha-num characters. Used to match VARCHAR type only partition keys
+     */
+    private static final String VARCHAR_OR_STRING_REGEX = "([a-zA-Z0-9_-\\s]+)";
+
+    private PartitionUtil()
+    {
+    }
+
+    /**
+     * Return a list of {@link AbstractMap.SimpleImmutableEntry<String, String>} instances
+     *
+     * @param table response of get table from AWS Glue
+     * @param partitionFolder      partition folder name
+     * @return List of {@link AbstractMap.SimpleImmutableEntry<String, String>} instances (column name, and value)
+     */
+    public static List<AbstractMap.SimpleImmutableEntry<String, String>> getPartitionColumnData(Table table, String partitionFolder)
+    {
+        Optional<String> optionalFolderRegex = getRegExExpression(table);
+        if (optionalFolderRegex.isPresent()) {
+            String folderRegex = optionalFolderRegex.get();
+            Pattern folderMatchPattern = Pattern.compile(folderRegex);
+            if (folderMatchPattern.matcher(partitionFolder).matches()) {
+                return getStoragePartitions(table.getParameters().get(PARTITION_PATTERN_KEY),
+                        partitionFolder, folderRegex, table.getPartitionKeys());
+            }
+        }
+        return List.of();
+    }
```

Comment:
Since you didn't understand my other comment as I mentioned, let me just rewrite the code to show you what I mean:

```
    public static List<AbstractMap.SimpleImmutableEntry<String, String>> getPartitionColumnData(Table table, String partitionFolder)
    {
        Optional<String> optionalFolderRegex = getRegExExpression(table);
        if (optionalFolderRegex.isPresent()) {
            String folderRegex = optionalFolderRegex.get();
            return getStoragePartitions(table.getParameters().get(PARTITION_PATTERN_KEY),
                partitionFolder, folderRegex, table.getPartitionKeys());
        }
        return List.of();
    }
```

Which can then be further simplified to:
```
    public static List<AbstractMap.SimpleImmutableEntry<String, String>> getPartitionColumnData(Table table, String partitionFolder)
    {
        getRegExExpression(table).stream().flatMap(folderRegex ->
            getStoragePartitions(table.getParameters().get(PARTITION_PATTERN_KEY),
                partitionFolder, folderRegex, table.getPartitionKeys());
        ).collect(Collectors.toList());
    }
```

Reply:
Let me know if I am missing something or if we can replace this code with the simplified version above.

Reply:
There was some glitch in the last refactored code, so took the first one.

Reply:
I implemented my first suggestion here: https://github.com/trianz-azam/aws-athena-query-federation/commit/46667987ec623399c16825c986c420328c92e0b2

Let me know if there's any problems with it.